cdc/redo/writer/file.go

```
@@ -0,0 +1,533 @@
+}
+
+func (w *Writer) renameInS3(ctx context.Context, oldPath, newPath string) error {
+	err := w.writeToS3(ctx, newPath)
+	if err != nil {
+		return cerror.WrapError(cerror.ErrS3StorageAPI, err)
+	}
+	return cerror.WrapError(cerror.ErrS3StorageAPI, w.storage.DeleteFile(ctx, filepath.Base(oldPath)))
+}
+
+func (w *Writer) getLogFileName() string {
+	if w.op != nil && w.op.getLogFileName != nil {
+		return w.op.getLogFileName()
+	}
+	return fmt.Sprintf("%s_%s_%d_%s_%d%s", w.cfg.CaptureID, w.cfg.ChangeFeedID, w.cfg.CreateTime.Unix(), w.cfg.FileType, w.commitTS.Load(), common.LogEXT)
+}
+
+func (w *Writer) filePath() string {
+	return filepath.Join(w.cfg.Dir, w.getLogFileName())
+}
+
+func openTruncFile(name string) (*os.File, error) {
+	return os.OpenFile(name, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, common.DefaultFileMode)
+}
+
+func (w *Writer) openNew() error {
+	err := os.MkdirAll(w.cfg.Dir, common.DefaultDirMode)
+	if err != nil {
+		return cerror.WrapError(cerror.ErrRedoFileOp, errors.Annotatef(err, "can't make dir: %s for new redo logfile", w.cfg.Dir))
+	}
+
+	// reset ts used in file name when new file
+	w.commitTS.Store(w.eventCommitTS.Load())
+	w.maxCommitTS.Store(w.eventCommitTS.Load())
+	path := w.filePath() + common.TmpEXT
+	f, err := openTruncFile(path)
+	if err != nil {
+		return cerror.WrapError(cerror.ErrRedoFileOp, errors.Annotate(err, "can't open new redo logfile"))
+	}
+	w.file = f
+	w.size = 0
+	err = w.newPageWriter()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (w *Writer) openOrNew(writeLen int) error {
+	path := w.filePath()
+	info, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		return w.openNew()
+	}
+	if err != nil {
+		return cerror.WrapError(cerror.ErrRedoFileOp, errors.Annotate(err, "error getting log file info"))
+	}
+
+	if info.Size()+int64(writeLen) >= w.cfg.MaxLogSize {
+		return w.rotate()
+	}
+
+	file, err := os.OpenFile(path, os.O_APPEND|os.O_WRONLY, common.DefaultFileMode)
+	if err != nil {
+		return w.openNew()
+	}
```

The file exists, but open it with failure, why we create a new file here, instead of returning an error.